contracts/multisig/contracts/src/AssetProxyOwner.sol

```diff
@@ -16,93 +16,205 @@
+    /// @param functionSelector 4 byte selector of registered function.
+    /// @param destination Address of destination where function will be called.
+    /// @param newSecondsTimeLocked Duration in seconds needed after a transaction is confirmed to become executable.
+    function registerFunctionCall(
+        bool hasCustomTimeLock,
+        bytes4 functionSelector,
+        address destination,
+        uint128 newSecondsTimeLocked
+    )
+        external
         onlyWallet
-        notNull(assetProxyContract)
     {
-        isAssetProxyRegistered[assetProxyContract] = isRegistered;
-        emit AssetProxyRegistration(assetProxyContract, isRegistered);
+        _registerFunctionCall(
+            hasCustomTimeLock,
+            functionSelector,
+            destination,
+            newSecondsTimeLocked
+        );
     }
 
-    /// @dev Allows execution of `removeAuthorizedAddressAtIndex` without time lock.
+    /// @dev Allows anyone to execute a confirmed transaction.
+    ///      Transactions *must* encode the values with the signature "bytes[] data, address[] destinations, uint256[] values"
+    ///      The `destination` and `value` fields of the transaction in storage are ignored.
+    ///      All function calls must be successful or the entire call will revert.
     /// @param transactionId Transaction ID.
-    function executeRemoveAuthorizedAddressAtIndex(uint256 transactionId)
+    function executeTransaction(uint256 transactionId)
         public
         notExecuted(transactionId)
         fullyConfirmed(transactionId)
-        validRemoveAuthorizedAddressAtIndexTx(transactionId)
     {
-        Transaction storage txn = transactions[transactionId];
-        txn.executed = true;
-        if (_externalCall(txn.destination, txn.value, txn.data.length, txn.data)) {
-            emit Execution(transactionId);
+        Transaction storage transaction = transactions[transactionId];
+        transaction.executed = true;
+
+        // Decode batch transaction data from transaction.data
+        // `destination` and `value` fields of transaction are ignored
+        // Note that `destination` must be non-0, or the transaction cannot be submitted
+        // solhint-disable
+        (
+            bytes[] memory data,
+            address[] memory destinations,
+            uint256[] memory values
+        ) = abi.decode(
+            transaction.data,
+            (bytes[], address[], uint256[])
+        );
+        // solhint-enable
+
+        // Ensure lengths of array properties are equal
+        uint256 length = data.length;
+        require(
+            length == destinations.length && length == values.length,
+            "EQUAL_LENGTHS_REQUIRED"
+        );
+
+        uint256 transactionConfirmationTime = confirmationTimes[transactionId];
+        for (uint i = 0; i != length; i++) {
+            // Ensure that each function call is past its timelock
+            _assertValidFunctionCall(
```

Comment:
It would be valuable to split this into two `for-loops`. The first would verify that all of the function call's are valid, and the second would actually make all of the calls. While this may slightly decrease efficiency when all of the functions have elapsed their timelocks, this would save a lot of gas in the case that they haven't.

Reply:
Discussed offline and decided to keep as is for max flexibility.